test/e2e/storage/testsuites/multivolume.go

```diff
@@ -521,17 +523,17 @@ func TestConcurrentAccessToSingleVolume(f *framework.Framework, cs clientset.Int
 		if i != 0 {
 			ginkgo.By(fmt.Sprintf("From pod%d, checking if reading the data that pod%d write works properly", index, index-1))
 			// For 1st pod, no one has written data yet, so pass the read check
-			utils.CheckReadFromPath(f, pod, *pvc.Spec.VolumeMode, path, byteLen, seed)
+			utils.CheckReadFromPath(f, pod, *pvc.Spec.VolumeMode, true, path, byteLen, seed)
 		}
 
 		// Update the seed and check if write/read works properly
 		seed = time.Now().UTC().UnixNano()
 
 		ginkgo.By(fmt.Sprintf("Checking if write to the volume in pod%d works properly", index))
-		utils.CheckWriteToPath(f, pod, *pvc.Spec.VolumeMode, path, byteLen, seed)
+		utils.CheckWriteToPath(f, pod, *pvc.Spec.VolumeMode, true, path, byteLen, seed)
 
```

Comment:
I might not completely understand the issue, but can it be resolved by just adding "sync command" here and the similar code in recheck?

Reply:
Maybe a real `sync` is sufficient, but the command that is part of BusyBox is not (no error, not synced). I tried it with a CentOS container, and there permissions prevented it from working.

Reply:
`sync` flushes pending writes to the disc, AFAICT it won't update the the page cache (at least on Linux).

Ideally we'd just drop the page cache in the "reading" node by doing `echo 3 > /proc/sys/vm/drop_caches `) before calling `dd`, but I don't think that's possible inside a container.

Reply:
Thank you for your explanation and testing.

How about `conv=fsync` option of dd command?

The dd command in busybox seems to support it and https://unix.stackexchange.com/questions/473854/block-device-cache-v-s-a-filesystem and https://stackoverflow.com/questions/49489798/is-running-sync-necessary-after-writing-a-disk-image seem to say that it will work well.

Reply:
I tried that as well. Unfortunately it does not help with the reading part where some readahead seems to happen (detecting partition table?). Because reading needs the whole sector for direct-io, the same amount of bytes need to be written (to make the checksum match).

Reply:
Now I understand that we need a way to avoid reader from reading from the cache, not just to guarantee the writer to write to the actual device before reader starts to read.
And it seems that the only way to achieve it in container is to use direct I/O, so I agree to use direct I/O for this purpose.

Reply:
> Reader side's requirement for avoid reading from cache needs to be resolved by using `iflag=nocache` in reading, however does writer side really needs direct I/O?

Unfortunately, there is no `iflag=nocache` option, the `nocache` is only for the `oflag` used by the writer.

You are correct, the writer does not require direct I/O. That side only needs to flush/sync the data after writing. The `oflag=nocache` can be replaced by `oflag=fdatasync` or similar. However, I do expect that applications using RWX block-mode PVCs use direct I/O on both sides when they need the data available on all participating nodes (like QEMU live-migration).

> 512 byte restriction seems to come from using `iflag=direct`, so I'm thinking about avoiding it by using other options in writing, like `conv=fsync` and/or `oflag=nocache` if we can really sync by these options.
> (Using debian image would be needed to use `iflag=nocache`, but avoiding direct I/O might still be possible. And it will potentially remove the TODOs in your codes.)
> 
> Does it make sense?

It really does, and you understood the issue perfectly! However, the missing support for `iflag=nocache` is preventing this from being an option. So the `iflag=direct` is the only alternative that is left. Direct I/O is the only option for readers to open a file/device without caching, see `man 2 open` and the `O_DIRECT` description, all the sync-flags for `open()` are related to writing only.

Reply:
> Direct I/O is the only option for readers to open a file/device without caching, see man 2 open and the O_DIRECT description, all the sync-flags for open() are related to writing only.

This can be confirmed by running this little program in pods allocated in different nodes (and with a shared RWX volume): https://gist.github.com/bertinatto/6783734c636907e699d999b7a99bd622

With the O_DIRECT flag it prints the same output regardless the node where the container runs. Without the flag, the results are different for each node.

Reply:
@nixpanic @bertinatto

Thank you for your explanation.

I understand that even though `iflag=nocache` can be passed to dd command, but won't be sufficient to skip cache.
So, I agree to apply this combination of options.